verilog/tools/kythe/indexing_facts_tree.h

```diff
@@ -0,0 +1,127 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_TOOLS_KYTHE_INDEXING_FACTS_TREE_H_
+#define VERIBLE_VERILOG_TOOLS_KYTHE_INDEXING_FACTS_TREE_H_
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <sstream>  // IWYU pragma: keep  // for ostringstream
+#include <string>   // for string, allocator, etc
+#include <utility>
+#include <vector>
+
+#include "absl/strings/substitute.h"
+#include "common/text/token_info.h"
+#include "verilog/tools/kythe/verilog_extractor_indexing_fact_type.h"
+
+// Anchor class represents the location and value of some token.
+class Anchor {
+ public:
+  Anchor(const verible::TokenInfo& token, absl::string_view base)
+      : start_location_(token.left(base)),
+        end_location_(token.right(base)),
+        value_(token.text()){};
+
+  // This function is for debugging only and isn't intended to be a textual
+  // representation of this class.
+  std::string DebugString() {
+    return absl::Substitute(
+        R"({"StartLocation": $0,"EndLocation": $1,"Value": "$2"})",
+        start_location_, end_location_, value_);
+  }
+
+ private:
+  // Start and end locations of the current token inside the code text.
+  int start_location_, end_location_;
+
+  // Value of the current token.
+  absl::string_view value_;
+};
+
+// This class is a simplified representation of CST and contains information
+// that can be used for extracting indexing-facts for different indexing tools.
+//
+// This is intended to be an abstract layer between the parser and the indexing
+// tool.
+class IndexingFactNode {
+ public:
+  explicit IndexingFactNode(IndexingFactType language_feature)
+      : indexing_fact_type_(language_feature) {}
+
+  IndexingFactNode(std::vector<Anchor> anchor,
+                   IndexingFactType language_feature)
+      : anchors_(std::move(anchor)), indexing_fact_type_(language_feature) {}
+
+  std::vector<IndexingFactNode> Children() { return children_; };
+
+  void AppendChild(IndexingFactNode indexing_fact_node) {
+    children_.push_back(std::move(indexing_fact_node));
+  };
+
+  void AppendChild(std::vector<IndexingFactNode> children) {
+    children_.insert(children_.end(), children.begin(), children.end());
+  };
+
+  void AppendAnchor(Anchor anchor) { anchors_.push_back(std::move(anchor)); };
+
+  void AppendAnchor(std::vector<Anchor> anchors) {
+    anchors_.insert(anchors_.end(), anchors.begin(), anchors.end());
+  };
+
+  // This function is for debugging only and isn't intended to be textual
+  // representation of this class.
+  std::string DebugString() {
+    std::string debug_string;
+    debug_string = R"({"IndexingFactType": ")" +
+                   LanguageFeatureEnumToString(indexing_fact_type_) + "\",\n";
+
+    debug_string += "\"Anchor\": [\n";
+    for (auto& anchor : anchors_) {
```

Comment:
You can simplify the code using absl::StrJoin(achors_, ",\n", DebugFormatter())
Where the debug formatter calls absl::StrAppend(out, t.DebugString());

More on strjoin, formatters and strappend at: https://abseil.io/docs/cpp/guides/strings